blockstore/splitstore/splitstore_prune.go

```diff
@@ -0,0 +1,592 @@
+// the given predicate.
+// missing references are ignored, as we expect to have plenty for snapshot syncs.
+func (s *SplitStore) walkChainDeep(ts *types.TipSet, retainStateP func(int64) bool,
+	f func(cid.Cid) error) error {
+	visited := cid.NewSet()
+	toWalk := ts.Cids()
+	walkCnt := 0
+
+	workers := runtime.NumCPU() / 2
+	if workers < 2 {
+		workers = 2
+	}
+
+	var wg sync.WaitGroup
+	workch := make(chan cid.Cid, 16*workers)
+	errch := make(chan error, workers)
+
+	var once sync.Once
+	defer once.Do(func() { close(workch) })
+
+	push := func(c cid.Cid) error {
+		if !visited.Visit(c) {
+			return nil
+		}
+
+		select {
+		case workch <- c:
+			return nil
+		case err := <-errch:
+			return err
+		}
+	}
+
+	worker := func() {
+		defer wg.Done()
+		for c := range workch {
+			err := s.walkObjectLax(c, f)
+			if err != nil {
+				errch <- xerrors.Errorf("error walking object (cid: %s): %w", c, err)
+				return
+			}
+		}
+	}
+
+	for i := 0; i < workers; i++ {
+		wg.Add(1)
+		go worker()
+	}
+
+	baseEpoch := ts.Height()
+	minEpoch := baseEpoch // for progress report
+	log.Infof("walking at epoch %d", minEpoch)
+
+	walkBlock := func(c cid.Cid) error {
+		if !visited.Visit(c) {
+			return nil
+		}
+
+		walkCnt++
+
+		if err := f(c); err != nil {
+			return err
+		}
+
+		var hdr types.BlockHeader
+		err := s.view(c, func(data []byte) error {
+			return hdr.UnmarshalCBOR(bytes.NewBuffer(data))
+		})
+
+		if err != nil {
+			return xerrors.Errorf("error unmarshaling block header (cid: %s): %w", c, err)
+		}
+
+		if hdr.Height < minEpoch {
+			minEpoch = hdr.Height
+			if minEpoch%10_000 == 0 {
+				log.Infof("walking at epoch %d (walked: %d)", minEpoch, walkCnt)
+			}
+		}
+
+		depth := int64(baseEpoch - hdr.Height)
+		retainState := retainStateP(depth)
+
+		if hdr.Height > 0 {
+			if err := push(hdr.Messages); err != nil {
+				return err
+			}
+			if retainState {
+				if err := push(hdr.ParentMessageReceipts); err != nil {
+					return err
+				}
+			}
+		}
+
+		if retainState || hdr.Height == 0 {
+			if err := push(hdr.ParentStateRoot); err != nil {
+				return err
+			}
+		}
+
+		if hdr.Height > 0 {
+			toWalk = append(toWalk, hdr.Parents...)
+		}
+
+		return nil
+	}
+
+	for len(toWalk) > 0 {
+		// walking can take a while, so check this with every opportunity
+		if err := s.checkClosing(); err != nil {
+			return err
+		}
+
+		select {
+		case err := <-errch:
+			return err
+		default:
+		}
+
+		walking := toWalk
+		toWalk = nil
+		for _, c := range walking {
+			if err := walkBlock(c); err != nil {
+				return xerrors.Errorf("error walking block (cid: %s): %w", c, err)
+			}
+		}
+	}
+
+	once.Do(func() { close(workch) })
+	wg.Wait()
+	select {
+	case err := <-errch:
+		return err
+	default:
+	}
+
+	log.Infow("chain walk done", "walked", walkCnt)
+
+	return nil
+}
+
+// like walkObject but treats missing references laxly; faster version of walkObjectIncomplete
+// without an occurs check.
```

Comment:
Is skipping the check actually faster here? I'd guess yes for receipts/messages, but not for stateroots which are mostly the same between epochs

Reply:
This is actually code I copied wholesale from @vyzo's original PR and do not understand.  It looks like this is ok as the problem is just that it might not be as fast as it claims it is.  But let me know if you think there is a problem here and I can explore